File: sdks/opencode-remote/android/app/src/main/java/com/opencode/remote/data/Protocol.kt
package com.opencode.remote.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

// ── 基础信封 ──

@Serializable
data class WsEnvelope(
    val seq: Int,
    val ts: Long,
    val payload: JsonObject
)

@Serializable
data class WsAck(val ack: Int)

@Serializable
data class AuthHandshake(
    val type: String = "auth",
    val role: String = "phone",
    val token: String
)

@Serializable
data class AuthResult(
    val type: String,
    val ok: Boolean,
    val error: String? = null
)

// ── 事件数据 ──

@Serializable
data class PermissionData(
    val id: String,
    val kind: String,
    val sessionID: String,
    val messageID: String,
    val callID: String? = null,
    val title: String,
    val metadata: JsonObject = JsonObject(emptyMap()),
    val created: Long
)

@Serializable
data class PermissionRepliedData(
    val sessionID: String,
    val permissionID: String,
    val response: String
)

@Serializable
data class SessionStatusData(
    val sessionID: String,
    val status: SessionStatus
)

@Serializable
data class SessionStatus(
    val type: String,
    val attempt: Int? = null,
    val message: String? = null,
    val next: Long? = null
)

@Serializable
data class SessionInfoData(
    val id: String,
    val title: String,
    val created: Long,
    val updated: Long
)

@Serializable
data class SessionErrorData(
    val sessionID: String? = null,
    val error: ErrorInfo? = null
)

@Serializable
data class ErrorInfo(
    val name: String,
    val message: String? = null
)

@Serializable
data class MessageDeltaData(
    val sessionID: String,
    val messageID: String,
    val partID: String,
    val partType: String,
    val delta: String? = null,
    val text: String? = null
)

@Serializable
data class ToolUpdateData(
    val sessionID: String,
    val messageID: String,
    val partID: String,
    val tool: String,
    val callID: String,
    val status: String,
    val title: String? = null,
    val input: JsonObject? = null,
    val output: String? = null,
    val error: String? = null
)

@Serializable
data class TodoItem(
    val content: String,
    val status: String,
    val priority: String
)

@Serializable
data class TodoUpdateData(
    val sessionID: String,
    val todos: List<TodoItem>
)

@Serializable
data class InstanceInfoSession(
    val id: String,
    val title: String,
    val status: String
)

@Serializable
data class InstanceInfoData(
    val instanceId: String = "",
    val version: String,
    val project: String,
    val directory: String,
    val sessions: List<InstanceInfoSession>
)

@Serializable
data class InstanceDisconnectedData(
    val instanceId: String
)

@Serializable
data class InstanceSyncData(
    val instanceIds: List<String>
)

@Serializable
data class MessageInfoData(
    val sessionID: String,
    val messageID: String,
    val role: String,
    val cost: Double? = null,
    val tokens: TokenInfo? = null,
    val error: ErrorInfo? = null
)

@Serializable
data class TokenInfo(
    val input: Int,
    val output: Int,
    val reasoning: Int
)

// ── Question 事件数据 ──

@Serializable
data class QuestionOption(
    val label: String,
    val description: String
)

@Serializable
data class QuestionInfo(
    val question: String,
    val header: String,
    val options: List<QuestionOption>,
    val multiple: Boolean? = null,
    val custom: Boolean? = null
)

@Serializable
data class QuestionAskedData(
    val id: String,
    val sessionID: String,
    val questions: List<QuestionInfo>,
    val tool: QuestionToolRef? = null
)

@Serializable
data class QuestionToolRef(
    val messageID: String,
    val callID: String
)

@Serializable
data class QuestionRepliedData(
    val sessionID: String,
    val requestID: String,
    val answers: List<List<String>>
)

@Serializable
data class QuestionRejectedData(
    val sessionID: String,
    val requestID: String
)

// ── Action 错误反馈 ──

@Serializable
data class ActionErrorData(
    val sessionID: String? = null,
    val actionType: String,
    val error: String
)

// ── Provider/Model 列表 ──

@Serializable
data class ModelItem(
    val id: String,
    val name: String,
    val reasoning: Boolean = false,
    val context: Int = 0,
    val output: Int = 0
)

@Serializable
data class ProviderInfo(
    val id: String,
    val name: String,
    val connected: Boolean = false,
    val models: List<ModelItem> = emptyList()
)

@Serializable
data class ModelRef(
    val providerID: String,
    val modelID: String
)

// ── 操作（Phone → Relay → OpenCode） ──

@Serializable
data class PermissionReplyAction(
    val type: String = "action.permission.reply",
    val data: PermissionReplyData
)

@Serializable
data class PermissionReplyData(
    val sessionID: String,
    val permissionID: String,
    val response: String,
    val feedback: String? = null
)

@Serializable
data class SessionMessageAction(
    val type: String = "action.session.message",
    val data: SessionMessageData
)

@Serializable
data class SessionMessageData(
    val sessionID: String,
    val content: String,
    val agent: String? = null,
    val model: ModelRef? = null
)

@Serializable
data class SessionAbortAction(
    val type: String = "action.session.abort",
    val data: SessionAbortData
)

@Serializable
data class SessionAbortData(val sessionID: String)

@Serializable
data class SessionCreateAction(
    val type: String = "action.session.create",
    val data: SessionCreateData
)

@Serializable
data class SessionCreateData(val title: String? = null)

@Serializable
data class RefreshAction(
    val type: String = "action.refresh",
    val data: JsonObject = JsonObject(emptyMap())
)

@Serializable
data class QuestionReplyAction(
    val type: String = "action.question.reply",
    val data: QuestionReplyData
)

@Serializable
data class QuestionReplyData(
    val sessionID: String,
    val questionID: String,
    val answer: String
)

@Serializable
data class QuestionRejectAction(
    val type: String = "action.question.reject",
    val data: QuestionRejectData
)

@Serializable
data class QuestionRejectData(
    val sessionID: String,
    val questionID: String
)

// ── 命令相关 ──

@Serializable
data class CommandInfo(
    val name: String,
    val description: String? = null,
    val hints: List<String>? = null
)

@Serializable
data class SessionCommandAction(
    val type: String = "action.session.command",
    val data: SessionCommandData
)

@Serializable
data class SessionCommandData(
    val sessionID: String,
    val command: String,
    val arguments: String? = null,
    val agent: String? = null,
    val model: ModelRef? = null
)

File: sdks/opencode-remote/shared/protocol.ts
/**
 * OpenCode Remote 通信协议
 *
 * 定义 Plugin ↔ Relay Server ↔ Phone App 三方之间的 WebSocket 消息格式。
 * 所有消息统一用 WsEnvelope 封装，通过 seq/ack 保证可靠投递。
 */

// ============================================================
// 基础信封
// ============================================================

/** WebSocket 消息信封，所有 payload 都包裹在内 */
export type WsEnvelope = {
  /** 单调递增序列号，用于 ACK 确认和离线重传 */
  seq: number
  /** 时间戳（毫秒） */
  ts: number
  /** 实际载荷 */
  payload: WsPayload
}

/** ACK 消息，确认收到指定 seq 的消息 */
export type WsAck = {
  ack: number
}

/** 心跳消息 */
export type WsPing = { type: "ping" }
export type WsPong = { type: "pong" }

/** 所有可能的 payload 类型 */
export type WsPayload = RemoteEvent | RemoteAction | WsPing | WsPong

// ============================================================
// 事件（OpenCode → Relay → Phone）
// ============================================================

/** 权限请求事件 — 需要用户在手机上审批 */
export type PermissionEvent = {
  type: "event.permission"
  data: {
    id: string
    kind: string
    pattern?: string | string[]
    sessionID: string
    messageID: string
    callID?: string
    title: string
    metadata: Record<string, unknown>
    created: number
  }
}

/** 权限已回复事件 */
export type PermissionRepliedEvent = {
  type: "event.permission.replied"
  data: {
    sessionID: string
    permissionID: string
    response: string
  }
}

/** 会话状态变更事件 */
export type SessionStatusEvent = {
  type: "event.session.status"
  data: {
    sessionID: string
    status: { type: "idle" } | { type: "busy" } | { type: "retry"; attempt: number; message: string; next: number }
  }
}

/** 会话创建/更新/删除事件 */
export type SessionInfoEvent = {
  type: "event.session.created" | "event.session.updated" | "event.session.deleted"
  data: {
    id: string
    title: string
    created: number
    updated: number
  }
}

/** 会话错误事件 */
export type SessionErrorEvent = {
  type: "event.session.error"
  data: {
    sessionID?: string
    error?: {
      name: string
      message?: string
    }
  }
}

/** AI 消息文本流增量 */
export type MessageDeltaEvent = {
  type: "event.message.delta"
  data: {
    sessionID: string
    messageID: string
    partID: string
    partType: string
    delta?: string
    text?: string
  }
}

/** 工具调用状态更新 */
export type ToolUpdateEvent = {
  type: "event.tool.update"
  data: {
    sessionID: string
    messageID: string
    partID: string
    tool: string
    callID: string
    status: "pending" | "running" | "completed" | "error"
    title?: string
    input?: Record<string, unknown>
    output?: string
    error?: string
    time?: { start?: number; end?: number }
  }
}

/** Todo 列表更新 */
export type TodoUpdateEvent = {
  type: "event.todo.updated"
  data: {
    sessionID: string
    todos: Array<{
      content: string
      status: string
      priority: string
    }>
  }
}

/** 实例信息 — 连接后首条消息，描述 OpenCode 实例基本信息 */
export type InstanceInfoEvent = {
  type: "event.instance.info"
  data: {
    instanceId: string
    version: string
    project: string
    directory: string
    sessions: Array<{
      id: string
      title: string
      status: "idle" | "busy" | "retry"
    }>
  }
}

/** 实例断开事件 — plugin 断开时 relay 通知 phone 移除对应终端 */
export type InstanceDisconnectedEvent = {
  type: "event.instance.disconnected"
  data: {
    instanceId: string
  }
}

/** 实例同步事件 — phone 连接时 relay 推送当前所有在线 plugin 列表 */
export type InstanceSyncEvent = {
  type: "event.instance.sync"
  data: {
    instanceIds: string[]
  }
}

/** 消息信息更新（含 token 用量、cost 等） */
export type MessageInfoEvent = {
  type: "event.message.info"
  data: {
    sessionID: string
    messageID: string
    role: "user" | "assistant"
    cost?: number
    tokens?: {
      input: number
      output: number
      reasoning: number
    }
    error?: {
      name: string
      message?: string
    }
  }
}

/** Agent 向用户提问事件 */
export type QuestionAskedEvent = {
  type: "event.question.asked"
  data: {
    id: string
    sessionID: string
    questions: Array<{
      question: string
      header: string
      options: Array<{
        label: string
        description: string
      }>
      multiple?: boolean
      custom?: boolean
    }>
    tool?: {
      messageID: string
      callID: string
    }
  }
}

/** 问题已回复事件 */
export type QuestionRepliedEvent = {
  type: "event.question.replied"
  data: {
    sessionID: string
    requestID: string
    answers: string[][]
  }
}

/** 问题已拒绝事件 */
export type QuestionRejectedEvent = {
  type: "event.question.rejected"
  data: {
    sessionID: string
    requestID: string
  }
}

/** Action 执行错误事件（Plugin → Phone 错误反馈） */
export type ActionErrorEvent = {
  type: "event.action.error"
  data: {
    sessionID?: string
    actionType: string
    error: string
  }
}

/** 可用模型列表事件 — Plugin 推送给 Phone */
export type ProviderListEvent = {
  type: "event.provider.list"
  data: {
    providers: Array<{
      id: string
      name: string
      connected: boolean
      models: Array<{
        id: string
        name: string
        reasoning: boolean
        context: number
        output: number
      }>
    }>
  }
}

/** 所有事件类型的联合 */
export type RemoteEvent =
  | PermissionEvent
  | PermissionRepliedEvent
  | SessionStatusEvent
  | SessionInfoEvent
  | SessionErrorEvent
  | MessageDeltaEvent
  | ToolUpdateEvent
  | TodoUpdateEvent
  | InstanceInfoEvent
  | MessageInfoEvent
  | QuestionAskedEvent
  | QuestionRepliedEvent
  | QuestionRejectedEvent
  | ActionErrorEvent
  | ProviderListEvent
  | InstanceDisconnectedEvent
  | InstanceSyncEvent
  | CommandListEvent

// ============================================================
// 操作（Phone → Relay → OpenCode）
// ============================================================

/** 回复权限请求 */
export type PermissionReplyAction = {
  type: "action.permission.reply"
  data: {
    sessionID: string
    permissionID: string
    /** once=仅本次，always=始终允许此模式，reject=拒绝 */
    response: "once" | "always" | "reject"
    feedback?: string
  }
}

/** 发送消息到会话 */
export type SessionMessageAction = {
  type: "action.session.message"
  data: {
    sessionID: string
    content: string
    agent?: string
    model?: {
      providerID: string
      modelID: string
    }
  }
}

/** 执行斜杠命令（如 /commit, /help 等） */
export type SessionCommandAction = {
  type: "action.session.command"
  data: {
    sessionID: string
    command: string
    arguments?: string
    agent?: string
    model?: {
      providerID: string
      modelID: string
    }
  }
}

/** 命令列表事件 — 连接时推送所有可用命令 */
export type CommandListEvent = {
  type: "event.command.list"
  data: {
    commands: Array<{
      name: string
      description?: string
      hints?: string[]
    }>
  }
}

/** 中止当前会话 */
export type SessionAbortAction = {
  type: "action.session.abort"
  data: {
    sessionID: string
  }
}

/** 创建新会话 */
export type SessionCreateAction = {
  type: "action.session.create"
  data: {
    title?: string
  }
}

/** 回复 agent 提出的问题 */
export type QuestionReplyAction = {
  type: "action.question.reply"
  data: {
    sessionID: string
    questionID: string
    answer: string
  }
}

/** 拒绝 agent 提出的问题 */
export type QuestionRejectAction = {
  type: "action.question.reject"
  data: {
    sessionID: string
    questionID: string
  }
}

/** 请求刷新实例信息 */
export type RefreshAction = {
  type: "action.refresh"
  data: Record<string, never>
}

/** 所有操作类型的联合 */
export type RemoteAction =
  | PermissionReplyAction
  | SessionMessageAction
  | SessionAbortAction
  | SessionCreateAction
  | QuestionReplyAction
  | QuestionRejectAction
  | RefreshAction
  | SessionCommandAction

// ============================================================
// 认证
// ============================================================

/** 连接握手时的认证消息（WebSocket 连接后首条消息） */
export type AuthHandshake = {
  type: "auth"
  role: "plugin" | "phone"
  token: string
  /** 仅 plugin 提供 */
  instance?: string
}

/** 认证响应 */
export type AuthResult = {
  type: "auth.result"
  ok: boolean
  error?: string
}

// ============================================================
// REST 降级接口的类型
// ============================================================

/** GET /api/events?since=<seq> 的响应 */
export type EventsResponse = {
  events: WsEnvelope[]
  latest: number
}

/** POST /api/action 的请求体 */
export type ActionRequest = {
  token: string
  action: RemoteAction
}
